components/TextFields/src/MDCTextInputController.m

```diff
@@ -0,0 +1,1199 @@
+      case UITextFieldViewModeAlways:
+        underlineColor = activeColor;
+        underlineHeight = MDCTextInputUnderlineActiveHeight;
+        break;
+      case UITextFieldViewModeWhileEditing:
+        underlineColor = self.textInput.isEditing ? activeColor : normalColor;
+        underlineHeight = self.textInput.isEditing ? MDCTextInputUnderlineActiveHeight
+                                                   : MDCTextInputUnderlineNormalHeight;
+        break;
+      case UITextFieldViewModeUnlessEditing:
+        underlineColor = !self.textInput.isEditing ? activeColor : normalColor;
+        underlineHeight = !self.textInput.isEditing ? MDCTextInputUnderlineActiveHeight
+                                                    : MDCTextInputUnderlineNormalHeight;
+        break;
+      case UITextFieldViewModeNever:
+      default:
+        underlineColor = normalColor;
+        underlineHeight = MDCTextInputUnderlineNormalHeight;
+        break;
+    }
+
+    self.textInput.underline.color =
+        (self.isDisplayingCharacterCountError || self.isDisplayingErrorText) ? self.errorColor
+                                                                             : underlineColor;
+    self.textInput.underline.lineHeight = underlineHeight;
+  }
+}
+
+#pragma mark - Underline Labels Fonts
+
++ (UIFont *)placeholderFont {
+  return [UIFont mdc_preferredFontForMaterialTextStyle:MDCFontTextStyleBody1];
+}
+
++ (UIFont *)underlineLabelsFont {
+  return [UIFont mdc_preferredFontForMaterialTextStyle:MDCFontTextStyleCaption];
+}
+
+#pragma mark - Properties Implementation
+
+- (void)setCharacterCountViewMode:(UITextFieldViewMode)characterCountViewMode {
+  if (_characterCountViewMode != characterCountViewMode) {
+    _characterCountViewMode = characterCountViewMode;
+
+    [self updateLayout];
+  }
+}
+
+- (void)setErrorAccessibilityValue:(NSString *)errorAccessibilityValue {
+  _errorAccessibilityValue = [errorAccessibilityValue copy];
+}
+
+- (void)setErrorColor:(UIColor *)errorColor {
+  if (![_errorColor isEqual:errorColor]) {
+    _errorColor = errorColor;
+    if (self.isDisplayingErrorText) {
+      [self updateLeadingUnderlineLabel];
+      [self updateTrailingUnderlineLabel];
+      [self updateUnderline];
+    }
+  }
+}
+
+- (void)setErrorText:(NSString *)errorText {
+  _errorText = [errorText copy];
+}
+
+- (void)setFloatingPlaceholderColor:(UIColor *)floatingPlaceholderColor {
+  if (![_floatingPlaceholderColor isEqual:floatingPlaceholderColor]) {
+    _floatingPlaceholderColor = floatingPlaceholderColor;
+    [self updatePlaceholder];
+  }
+}
+
+- (void)setFloatingPlaceholderScale:(NSNumber *)floatingPlaceholderScale {
+  if (![_floatingPlaceholderScale isEqualToNumber:floatingPlaceholderScale]) {
+    _floatingPlaceholderScale = floatingPlaceholderScale;
+    [self updatePlaceholder];
+  }
+}
+
+- (void)setHelperText:(NSString *)helperText {
+  if (self.isDisplayingErrorText) {
+    self.previousLeadingText = [helperText copy];
```

Comment:
You don't have to copy helperText here since previousLeadingText already copies.